src/agent.rs

```diff
@@ -0,0 +1,574 @@
+                        #[allow(clippy::mem_forget)] // later dropped by head
+                        mem::forget(m);
+                        let new_m = Mbuf::new_with_ptr(mo).ok()?;
+                        Some(new_m) // fragmented ip packet
+                    } else {
+                        Some(m) // unfragmented ip packet
+                    }
+                }?;
+                return if proto_id == IP_NEXT_PROTO_UDP {
+                    handle_ipv4_udp(m)
+                } else {
+                    debug!("Unrecognized proto id {proto_id}");
+                    None
+                };
+            }
+            RTE_ETHER_TYPE_IPV6 | RTE_ETHER_TYPE_ARP => {}
+            ether_type => error!("Unsupported ether type {ether_type:x}"),
+        }
+    }
+    None
+}
+
+#[allow(unsafe_code)]
+impl RxAgent {
+    /// Start an `RxAgent`, spawn a thread to do the polling job.
+    pub(crate) fn start(socket_id: i32) -> Arc<Self> {
+        let running = AtomicBool::new(true);
+        let this = Arc::new(RxAgent {
+            running,
+            tasks: Mutex::new(BTreeSet::new()),
+        });
+        let that = Arc::clone(&this);
+        let _handle = task::spawn_blocking(move || {
+            let mut frag_tbl = IpFragmentTable::new(socket_id)?;
+            let mut death_row = IpFragDeathRow::new(socket_id)?;
+
+            while that.running.load(Ordering::Acquire) {
+                let tasks = that.tasks.lock().map_err(Error::from)?;
+                let task_iter = tasks.iter();
+                for &(port_id, queue_id) in task_iter {
+                    let mut ptrs = vec![ptr::null_mut(); MAX_PKT_BURST as usize];
+                    // SAFETY: ffi
+                    let n = unsafe {
+                        rte_eth_rx_burst(port_id, queue_id, ptrs.as_mut_ptr(), MAX_PKT_BURST)
+                    };
+                    trace!("{n} packets received");
+                    for ptr in ptrs.into_iter().take(n as _) {
+                        let m = Mbuf::new_with_ptr(ptr)?;
+                        if let Some((sockfd, res)) = handle_ether(m, &mut frag_tbl, &mut death_row)
+                        {
+                            let res = socket::put_mailbox(sockfd, res);
+                            if let Err(e) = res {
+                                error!("An error {e} occurred in `put_mailbox`");
+                            }
+                        }
+                    }
+                }
+            }
+            info!("RxAgent thread terminated");
+            Result::<()>::Ok(())
+        });
+        this
+    }
+
+    /// Stop the `RxAgent`.
+    pub(crate) fn stop(self: &Arc<Self>) {
+        self.running.store(false, Ordering::Release);
+    }
+
+    /// Register a (`port_id`, `queue_id`) to an `RxAgent`.
+    ///
+    /// Adds a (`port_id`, `queue_id`) pair to the set to be polled.
+    pub(crate) fn register(self: &Arc<Self>, port_id: u16, queue_id: u16) -> Result<()> {
```

Comment:
Can we call `register` after `stop`ing it? If yes, would it cause any error? Is it still safe?